src/Microsoft.ML.Transforms/KeyToBinaryVectorTransform.cs

```diff
@@ -44,311 +76,395 @@ private static VersionInfo GetVersionInfo()
             // <base>
-            Host.AssertNonEmpty(Infos);
-
-            Init(out _concat, out _types, out _bitsPerKey);
+            ctx.CheckAtModel();
+            ctx.SetVersionInfo(GetVersionInfo());
+            SaveColumns(ctx);
         }
 
-        private void Init(out bool[] concat, out VectorType[] types, out int[] bitsPerKey)
+        // Factory method for SignatureLoadModel.
+        public static KeyToBinaryVectorTransform Create(IHostEnvironment env, ModelLoadContext ctx)
         {
-            concat = new bool[Infos.Length];
-            types = new VectorType[Infos.Length];
-            bitsPerKey = new int[Infos.Length];
+            Contracts.CheckValue(env, nameof(env));
+            var host = env.Register(RegistrationName);
 
-            for (int i = 0; i < Infos.Length; i++)
-                ComputeType(this, Source.Schema, i, Infos[i], Metadata,
-                    out _types[i], out _concat[i], out _bitsPerKey[i]);
+            host.CheckValue(ctx, nameof(ctx));
+            ctx.CheckAtModel(GetVersionInfo());
 
-            Metadata.Seal();
+            return new KeyToBinaryVectorTransform(host, ctx);
         }
 
-        public static KeyToBinaryVectorTransform Create(IHostEnvironment env, ModelLoadContext ctx, IDataView input)
+        private KeyToBinaryVectorTransform(IHost host, ModelLoadContext ctx)
+            : base(host, ctx)
         {
-            Contracts.CheckValue(env, nameof(env));
-            var h = env.Register(RegistrationName);
-            h.CheckValue(ctx, nameof(ctx));
-            h.CheckValue(input, nameof(input));
-            return h.Apply("Loading Model", ch => new KeyToBinaryVectorTransform(h, ctx, input));
         }
 
-        public override void Save(ModelSaveContext ctx)
+        public static IDataTransform Create(IHostEnvironment env, IDataView input, params ColumnInfo[] columns) =>
+            new KeyToBinaryVectorTransform(env, input, columns).MakeDataTransform(input);
+
+        // Factory method for SignatureDataTransform.
+        public static IDataTransform Create(IHostEnvironment env, Arguments args, IDataView input)
         {
-            Host.CheckValue(ctx, nameof(ctx));
+            Contracts.CheckValue(env, nameof(env));
+            env.CheckValue(args, nameof(args));
+            env.CheckValue(input, nameof(input));
 
-            // *** Binary format ***
-            // <prefix handled in static Create method>
-            // <base>
-            ctx.CheckAtModel();
-            ctx.SetVersionInfo(GetVersionInfo());
-            SaveBase(ctx);
+            env.CheckValue(args.Column, nameof(args.Column));
+            var cols = new ColumnInfo[args.Column.Length];
+            using (var ch = env.Start("ValidateArgs"))
+            {
+                for (int i = 0; i < cols.Length; i++)
+                {
+                    var item = args.Column[i];
+                    cols[i] = new ColumnInfo(item.Source, item.Name);
```

Comment:
>item.Source [](start = 45, length = 11)

`item.Source ?? item.Name`

Reply:
Why, don't we validate all that stuff during parsing?

---
In reply to: [216461891](https://github.com/dotnet/machinelearning/pull/858#discussion_r216461891) [](ancestors = 216461891)

Reply:
for args, it's valid to have `Source=null`, for column info's it's not

---
In reply to: [216482894](https://github.com/dotnet/machinelearning/pull/858#discussion_r216482894) [](ancestors = 216482894,216461891)